pyupgrade/_data.py

```diff
@@ -103,6 +104,17 @@ def visit(
                 next_state = state._replace(in_annotation=True)
             else:
                 next_state = state
+            if isinstance(node, ast.Subscript):
+                if isinstance(node.value, ast.Name):
+                    if node.value.id in {
+                        'Literal',
+                        'TypedDict',
+                        'NamedTuple',
+                        'Annotated',
+                    }:
+                        next_state = (
+                            next_state._replace(exclude_from_pep563=True)
+                        )
```

Comment:
ah, you don't need to do this, I just meant implement your helper function non-recursively (use a local `list` / `deque` as a stack / queue)

Reply:
Thanks, I'll look into how to do that 

Though isn't this approach simpler, as it means only overriding Constant/Str rather than having to override FunctionDef and AnnAssign and then doing complicatedgymnastics to go through their attributes? The previous solution felt overly complex

Reply:
continually rebuilding the state object is very slow -- and adding specific state for a single plugin is not good